src/utils.jl

```diff
@@ -29,3 +29,69 @@ function compute_discontinuity_tree(lags, alg, start_val)
     start_val + unique(vcat((sum.(collect(with_replacement_combinations(lags, i))) for i in
                              1:alg_order(alg))...))
 end
+
+"""
+    reduce_solution!(integrator::DDEIntegrator, tmax)
+
+Remove time points of ODE solution of `integrator` up to time `tmax` that are not required
+for calculation of DDE solution.
+"""
+@inline function reduce_solution!(integrator::DDEIntegrator, tmax)
+    if !integrator.opts.dense && !integrator.opts.save_everystep &&
+        integrator.saveiter < integrator.integrator.saveiter
+
+        # last time point of ODE solution that is known to be required for DDE solution
+        @inbounds t_sol_prev = integrator.integrator.sol.t[integrator.saveiter]
+        # next time point of ODE solution that might be required for DDE solution
+        @inbounds t_sol = integrator.integrator.sol.t[integrator.saveiter + 1]
+
+        @inbounds while t_sol < tmax && integrator.saveiter < integrator.integrator.saveiter
+            needed = false
+
+            while !(isempty(integrator.saveat)) && top(integrator.saveat) <= t_sol
+                # do not remove time point if it is upper bound of a required
+                # interpolation interval
+                if t_sol_prev < top(integrator.saveat)
+                    needed = true
+                end
+
+                # remove passed time points
+                pop!(integrator.saveat)
+            end
+
+            # do not remove time point if it is lower bound of a required interpolation
+            # interval
+            needed = needed || (!(isempty(integrator.saveat)) &&
+                                integrator.saveiter + 1 < integrator.integrator.saveiter &&
+                                t_sol < top(integrator.saveat) <
+                                integrator.integrator.sol.t[integrator.saveiter + 2])
+
+            if !needed
+                # delete not required time point, function value, and interpolation data
+                deleteat!(integrator.integrator.sol.t, integrator.saveiter + 1)
```

Comment:
It's not necessary to delete these entries, but I wanted to check if there is any benefit in deleting unneeded entries already during the algorithm.
The problem is that the entries that can be deleted do not have to be successive entries - some entries might be needed to calculate interpolated values of the output solution whereas others can be removed completely... Maybe one could completely remove these entries and only keep the interpolated values - but I do not know if this could cause other problems....